src/main/java/com/axonivy/market/service/impl/ProductServiceImpl.java

```diff
@@ -0,0 +1,251 @@
+import com.axonivy.market.github.service.GHAxonIvyMarketRepoService;
+import com.axonivy.market.github.service.GithubService;
+import com.axonivy.market.github.util.GithubUtils;
+import com.axonivy.market.repository.GithubRepoMetaRepository;
+import com.axonivy.market.repository.ProductRepository;
+import com.axonivy.market.service.ProductService;
+
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+@Service
+public class ProductServiceImpl implements ProductService {
+
+  private final ProductRepository productRepository;
+  private final GHAxonIvyMarketRepoService axonivyMarketRepoService;
+  private final GithubRepoMetaRepository githubRepoMetaRepository;
+  private final GithubService githubService;
+
+  private GHCommit lastGHCommit;
+  private GithubRepoMeta marketRepoMeta;
+
+  public ProductServiceImpl(ProductRepository productRepository, GHAxonIvyMarketRepoService axonivyMarketRepoService,
+      GithubRepoMetaRepository githubRepoMetaRepository, GithubService githubService) {
+    this.productRepository = productRepository;
+    this.axonivyMarketRepoService = axonivyMarketRepoService;
+    this.githubRepoMetaRepository = githubRepoMetaRepository;
+    this.githubService = githubService;
+  }
+
+  @Override
+  public Page<Product> findProducts(String type, String keyword, Pageable pageable) {
+    final var filterType = FilterType.of(type);
+    if (StringUtils.isNoneBlank(keyword)) {
+      return searchProducts(filterType, keyword, pageable);
+    }
+
+    syncLatestDataFromMarketRepo();
+
+    Pageable unifiedPageabe = refinePagination(pageable);
+    return switch (filterType) {
+    case ALL -> productRepository.findAll(unifiedPageabe);
+    case CONNECTORS, UTILITIES, SOLUTIONS -> productRepository.findByType(filterType.getCode(), pageable);
+    default -> Page.empty();
+    };
+  }
+
+  @Override
+  public boolean syncLatestDataFromMarketRepo() {
+    var isAlreadyUpToDate = isLastGithubCommitCovered();
+    if (!isAlreadyUpToDate) {
+      if (marketRepoMeta == null) {
+        syncProductsFromGithubRepo();
+        marketRepoMeta = new GithubRepoMeta();
+      } else {
+        updateLatestChangeToProductsFromGithubRepo();
+      }
+      syncRepoMetaDataStatus();
+    }
+    return isAlreadyUpToDate;
+  }
+
+  private void syncRepoMetaDataStatus() {
+    if (marketRepoMeta == null || lastGHCommit == null) {
+      return;
+    }
+    String repoURL = Optional.ofNullable(lastGHCommit.getOwner()).map(GHRepository::getUrl).map(URL::getPath)
+        .orElse(EMPTY);
+    marketRepoMeta.setRepoURL(repoURL);
+    marketRepoMeta.setRepoName(GitHubConstants.AXONIVY_MARKETPLACE_REPO_NAME);
+    marketRepoMeta.setLastSHA1(lastGHCommit.getSHA1());
+    marketRepoMeta.setLastChange(GithubUtils.getGHCommitDate(lastGHCommit));
+    githubRepoMetaRepository.save(marketRepoMeta);
+    marketRepoMeta = null;
+  }
+
+  private void updateLatestChangeToProductsFromGithubRepo() {
+    if (lastGHCommit == null || marketRepoMeta == null) {
```

Comment:
MarketRepoMeta is always not null. Do we need the condition `marketRepoMeta == null`?